### DISCUSSION TITLE
Local Defs

### DISCUSSION THREAD

**Discussion by magnus-madsen** [flix/flix#7780] [Ideas]
This incomplete program:

```scala
def main(): Unit \ IO = 
    let x = 123;

def inc(x: Int32): Int32 = x + 1
```

Suffers from the fact that `inc` is believed to be a function local to `main` when it is actually meant to be a top-level function.

This breaks IDE tooling.

This happens because we use the same keyword (`def`) for top-level and local defs.

We can fix this in a few ways:

- We can use a different keyword for local `def`s:
  - e.g. we could write `let foo(a, b, c) = ...`. 
  - Other proposed keywords include `local def` or `let def`.
- We could _require_ curly braces around functions that have statements/lets/defs. (like Scala).
- We can also leave it be. Notably, if `inc` was declared as `pub def` the problem would go away.

**Comment by magnus-madsen** [comment]
We -- @JonathanStarup @AStenbaek @chanattan discussed this yesterday. Do you @mlutze or @stephentetley have any opinions?

**Comment by magnus-madsen** [comment]
@herluf-ba Does the problem with `def` occur in other context? We do reuse some keywords (`with`) comes to mind, but I don't think they suffer from the same problem?

**Reply by herluf-ba** [reply]
There's no other examples of this as far as I see.

**Comment by herluf-ba** [comment]
I think the best option is `local def ...` / `let def ...`. Generally "prefixing with intent" works out well for resilient parsing.

**Comment by mlutze** [comment]
In general, my style is to include braces unless the function is a one-liner, so the Scala-like rules would work for me 

`local` keyword feels a little bad, but since the typing rules of local defs are slightly different (no generalization), maybe a larger difference is appropriate. Will we be able to have an error that says "local needed here"?

**Reply by herluf-ba** [reply]
Yea, in this example:
```scala
def length(l: List[a]): Int32 = {
    def loop(ll, acc) = match ll {
        case Nil     => acc
        case _ :: xs => loop(xs, acc + 1)
    };
    loop(l, 0)
}

def main(): Int32 = length(List#{ 1, 2, 3 })
```
We would be able to say:
```
>> Expected 'local' before 'def'

    def loop(ll, acc) = match ll {
    ^ Here
    
 Hint: prefix local definitions with 'local'.
```

The same error would apply in the example we have been discussing:
```scala
def main(): Unit \ IO = 
    let x = 123;

def inc(x: Int32): Int32 = x + 1
```
Would be:
```

>> Expected 'local' before 'def'

def inc(x: Int32): Int32 = x + 1
^ Here
    
 Hint: prefix local definitions with 'local'.
```
This at least indicates that `inc` is being read as a local def

**Comment by chanattan** [comment]
Why would `let x = 123;` be required to have ';' at the end?
It seemed to me in other languages that the last statement would not require a semicolon to explicit tell that it is the end statement, in particular the return statement.

**Reply by JonathanStarup** [reply]
Its not always clear where the next "thing" starts. here is a made up scala example just to illustrate
```
val x = 123;
-42;
println(x)
```
without `;`
```
val x = 123
-42
println(x)
```
Does it now print 123 or 81?

But yeah you shouldn't end your expression with `;`, but if you do so it might give unclear parsing errors

**Reply by chanattan** [reply]
I see, thank you.
I would say first that simple referencing or useless computed values such as 42 would not be allowed, unless as return values. I think this is the case in some language.

And I implicitly chose here to force semicolon except for the last statement, which should be the return value. This is how I felt it would be in functional languages. In this case, it should return 123 in all cases, otherwise there would be a syntax error.

I believe that functional programmers should know where their function ends with a return value, explicitly known with no semi-column at the end of their statement. Otherwise, anything below should be considered part of the function.
